vowpalwabbit/csv_parser/src/parse_example_csv.cc

```diff
@@ -0,0 +1,280 @@
+{
+  parse_label(all, ae, csv_line);
+  // TODO: parse_tag(all, ae, csv_line);
+  parse_namespaces(all, ae, csv_line);
+}
+
+void parser::parse_label(VW::workspace* all, VW::example* ae, std::vector<VW::string_view> csv_line)
+{
+  all->example_parser->lbl_parser.default_label(ae->l);
+
+  // negative numbers support the "from end of line" convention
+  // (e.g: -1 is last column, -2 is next-to-last, etc.)
+  if (all->label < 0) { label_index = csv_line.size() + all->label; }
+  else
+  {
+    label_index = all->label;
+  }
+
+  if (label_index >= csv_line.size() || label_index < 0) { THROW("Label index out of range!"); }
+
+  VW::string_view label_content(csv_line[label_index]);
+
+  // Multiclass labels will be auto-converted to 1..k if they are
+  // non-numeric e.g. Species: {setosa, versicolor, virginica} -> {1, 2, 3}
+  std::string label_string;
+  if (!check_if_float(label_content))
+  {
+    auto it = multiclass_label_counter.find(label_content);
+    if (it == multiclass_label_counter.end())
+    {
+      label_string = std::to_string(multiclass_label_counter.size() + 1);
+      multiclass_label_counter.insert(std::make_pair(label_content, multiclass_label_counter.size() + 1));
+    }
+    else
+    {
+      label_string = std::to_string(it->second);
+    }
+    label_content = VW::string_view(label_string);
+  }
+
+  all->example_parser->words.clear();
+  VW::tokenize(' ', label_content, all->example_parser->words);
+
+  if (!all->example_parser->words.empty())
+  {
+    all->example_parser->lbl_parser.parse_label(ae->l, ae->_reduction_features,
+        all->example_parser->parser_memory_to_reuse, all->sd->ldict.get(), all->example_parser->words, all->logger);
+  }
+}
+
+void parser::parse_namespaces(VW::workspace* all, example* ae, std::vector<VW::string_view> csv_line)
+{
+  // TODO: multiple namespaces support, currently using the filename as namespace
+  std::string ns = all->data_filename;
+  size_t found = ns.find_last_of('.');
+  if (found != std::string::npos) { ns = ns.substr(0, found); }
+
+  unsigned char _index = static_cast<unsigned char>(ns[0]);
+  bool new_index = false;
+
+  _channel_hash = all->example_parser->hasher(ns.data(), ns.length(), all->hash_seed);
+
+  if (ae->feature_space[_index].size() == 0) { new_index = true; }
+
+  ae->feature_space[_index].start_ns_extent(_channel_hash);
+  parse_features(all, ae->feature_space[_index], csv_line, (all->audit || all->hash_inv) ? ns.c_str() : nullptr);
+  if (new_index && ae->feature_space[_index].size() > 0) { ae->indices.push_back(_index); }
+
+  ae->feature_space[_index].end_ns_extent();
+}
+
+void parser::parse_features(VW::workspace* all, features& fs, std::vector<VW::string_view> csv_line, const char* ns)
+{
+  size_t _anon = 0;
+  for (size_t i = 0; i < csv_line.size(); i++)
+  {
+    // Skip label
+    if (label_index == i) { continue; }
+    // Ignore namespace info value
+    float _cur_channel_v = 1.f;
```

Comment:
This is an interesting one, we can't scale the namespace value without a way to provide that information. Using the pre-header header we could potentially place this information there

Reply:
Yeap, but currently I can't come up with a good way to represent that info while still maintaining csv's original representing format. So I'm just leave the _cur_channel_v here for future discussion. I guess having the namespace value feature is also not something emergent.

Reply:
Maybe specify the namespace info value from the command line is a good choice.

Reply:
In my last commit, I've added a new option to Scale the namespace values by specifying the float ratio. e.g. --csv_ns_value=a:0.5,b:0.3,:8

Reply:
We could add this to the CSV header we have been talking about [here](https://github.com/HollowMan6/vowpal_wabbit/pull/1#discussion_r870392165)?

Reply:
I guess it would be overly restrictive for

```csv
sepal | length, sepal | width, petal | length, petal | width
1,2,3,4
```

since it's the namespace info value, and the namespace can occur multiple times here (`sepal`, `petal`), it's not a good choice to have those values repeat multiple times.

```csv
sepal:2.1 | length, sepal:2.1 | width, petal:1.1 | length, petal:1.1 | width
1,2,3,4
```

Reply:
You're right, the namespace repetition makes this a poor place to put it. Okay, how about we leave out namespace scaling for now from the file format then? It is a rarely used feature anyway so I think this is okay